kolibri/plugins/management/assets/src/actions.js

```diff
@@ -69,110 +76,136 @@ function _managePageTitle(title) {
-    // oldRole is admin and role is coach or oldRole is coach and role is admin.
-      const OldRoleModel = RoleResource.getModel(oldRoldID);
-      OldRoleModel.delete().then(() => {
-      // create new role when old role is successfully deleted.
-        const rolePayload = {
-          user: id,
-          collection: FacilityUserModel.attributes.facility,
-          kind: role,
-        };
-        const RoleModel = RoleResource.createModel(rolePayload);
-        RoleModel.save(rolePayload).then((newRole) => {
-        // update the facilityUser when new role is successfully created.
-          FacilityUserModel.save(payload).then(responses => {
-            // force role change because if the role is the only changing attribute
-            // FacilityUserModel.save() will not send request to server.
-            responses.roles = [newRole];
-            store.dispatch('UPDATE_USERS', [responses]);
-          })
-          .catch(error => { coreActions.handleApiError(store, error); });
-        });
-      })
-      .catch(error => { coreActions.handleApiError(store, error); });
-    } else {
-    // role is learner and oldRole is admin or coach.
-      const OldRoleModel = RoleResource.getModel(oldRoldID);
-      OldRoleModel.delete().then(() => {
-        FacilityUserModel.save(payload).then(responses => {
-          // force role change because if the role is the only changing attribute
-          // FacilityUserModel.save() will not send request to server.
-          responses.roles = [];
-          store.dispatch('UPDATE_USERS', [responses]);
-        })
-        .catch(error => { coreActions.handleApiError(store, error); });
+function updateUser(store, stateUser) {
+  // payload needs username, fullname, and facility
+  const userID = stateUser.id;
+  const savedUserModel = FacilityUserResource.getModel(userID);
+  const savedUser = savedUserModel.attributes;
+  const changedValues = {};
+  let roleAssigned = Promise.resolve(savedUserModel.attributes);
+
+  // explicit checks for the only values that can be changed
+  if (stateUser.full_name !== savedUser.full_name) {
+    changedValues.full_name = stateUser.full_name;
+  }
+
+  if (stateUser.username !== savedUser.username) {
+    changedValues.username = stateUser.username;
+  }
+
+  if (stateUser.password !== savedUser.password) {
+    changedValues.password = stateUser.password;
+  }
+
+  if (stateUser.facility !== savedUser.facility) {
+    changedValues.facility = stateUser.facility;
+  }
+
+  if (stateUser.kind !== _userState(savedUser).kind) {
+    // assumes there's no previous roles to delete at first
+    let handlePreviousRoles = Promise.resolve();
+
+    if (savedUser.roles.length) {
+      const roleDeletes = [];
+      savedUser.roles.forEach(role => {
+        roleDeletes.push(RoleResource.getModel(role.id).delete());
       });
+
+
+      // delete the old role models if this was not a learner
+      handlePreviousRoles = Promise.all(roleDeletes).then(responses => {
```

Comment:
Yeah - it does feel a little weird to re-assign a new promise to `handlePreviousRoles` inside the `if` statement, but I understand what you're doing. It will work, but it's something to maybe avoid in the future.

The main issue is that the first promise is just a throw-away. When you look at the code carefully it becomes clear that's all right here, but in general it would be dangerous to just overwrite a promise like that.